tests/backends/data/test_clickhouse.py

```diff
@@ -0,0 +1,589 @@
+
+    results = list(backend.read(chunk_size=1000))
+    assert len(results) == 3
+    assert results[0]["event"] == statements[0]
+    assert results[1]["event"] == statements[1]
+    assert results[2]["event"] == statements[2]
+
+    results = list(backend.read(raw_output=True))
+    assert len(results) == 3
+    assert isinstance(results[0], bytes)
+    assert json.loads(results[0])["event"] == statements[0]
+
+
+# pylint: disable=unused-argument
+def test_backends_data_clickhouse_data_backend_read_method_with_a_custom_query(
+    clickhouse, clickhouse_backend
+):
+    """Test the `ClickHouseDataBackend.read` method with a custom query."""
+    date_1 = (datetime.now() - timedelta(seconds=3)).isoformat()
+    date_2 = (datetime.now() - timedelta(seconds=2)).isoformat()
+    date_3 = (datetime.now() - timedelta(seconds=1)).isoformat()
+
+    statements = [
+        {"id": str(uuid.uuid4()), "bool": 1, "timestamp": date_1},
+        {"id": str(uuid.uuid4()), "bool": 0, "timestamp": date_2},
+        {"id": str(uuid.uuid4()), "bool": 1, "timestamp": date_3},
+    ]
+
+    backend = clickhouse_backend()
+    documents = list(
+        backend._to_documents(statements)
+    )  # pylint: disable=protected-access
+    backend._bulk_import(documents)  # pylint: disable=protected-access
+
+    # Test filtering
+    query = ClickHouseQuery(where_clause="event.bool = 1")
+    results = list(backend.read(query=query, chunk_size=None))
+    assert len(results) == 2
+    assert results[0]["event"] == statements[0]
+    assert results[1]["event"] == statements[2]
+
+    # Test fields
+    query = ClickHouseQuery(return_fields=["event_id", "event.bool"])
+    results = list(backend.read(query=query))
+    assert len(results) == 3
+    assert len(results[0]) == 2
+    assert results[0]["event_id"] == documents[0][0]
+    assert results[0]["event.bool"] == statements[0]["bool"]
+    assert results[1]["event_id"] == documents[1][0]
+    assert results[1]["event.bool"] == statements[1]["bool"]
+    assert results[2]["event_id"] == documents[2][0]
+    assert results[2]["event.bool"] == statements[2]["bool"]
+
+    # Test filtering and projection
+    query = ClickHouseQuery(
+        where_clause="event.bool = 0", return_fields=["event_id", "event.bool"]
+    )
+    results = list(backend.read(query=query))
+    assert len(results) == 1
+    assert len(results[0]) == 2
+    assert results[0]["event_id"] == documents[1][0]
+    assert results[0]["event.bool"] == statements[1]["bool"]
+
+
+def test_backends_data_clickhouse_data_backend_list_method(
+    clickhouse, clickhouse_backend
+):
+    """Test the `ClickHouseDataBackend.list` method."""
+
+    backend = clickhouse_backend()
+
+    assert list(backend.list(details=True)) == [{"name": CLICKHOUSE_TEST_TABLE_NAME}]
+    assert list(backend.list(details=False)) == [CLICKHOUSE_TEST_TABLE_NAME]
+
+
+def test_backends_data_clickhouse_data_backend_to_documents_method_invalid_timestamp(
```

Comment:
Could we test the `write` method given documents with invalid timestamps instead? (to avoid testing private methods) WDYT?

Reply:
Indeed, have to test the `write` method to be less implementation-dependant :+1: